PR 3 of n: Add ML notebook on cluster validation

jaclyn-taroni opened the pull request:
Related to #244 and #257 

The notebook being added in #259 ends with the following question:

> How many groups of medulloblastoma samples do you think there are?

In the notebook being added here, we subset the OpenPBTA dataset to medulloblastoma samples only and cover the following:

* Consensus clustering
* Calculating silhouette coefficient (using labels from consensus clustering)
* Subsetting to high variance genes and then performing dimension reduction - see https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/730#issuecomment-660511043 (the part about the paper where the subtypes were initially described, specifically)
  * UMAP
  * PCA

What could use more context?

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
```diff
@@ -0,0 +1,326 @@
+# Like umap(), prcomp() expects the features to be columns
+pca_results <- prcomp(t(rnaseq_mat[high_variance_index, ]), 
+                      scale = TRUE)
+# The low-dimensional representation is returned in x
```
I was looking at this in the rendered notebook, so while this is not my full review, I noticed this comment mostly because the `head()` still prints a whole lot. Technically, PCA has the same dimensionality, we just throw some away after calculating them. `calculatePCA()` does this by default, so that may be what they expect, but you might want to mention that `prcomp()` does not (unless you set `tol`), so the output matrix has the same dimensions as the input. 

At least that is my understanding!

jashapiro replied:
In any case, you might want to use `pca_results[1:6, 1:6]` or something similar to reduce the printout.

jaclyn-taroni replied:
Sounds good 👍  `calculatePCA()` returns the first 50 PCs based on my read.

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
````diff
@@ -0,0 +1,351 @@
+[`plotly`](https://plotly.com/r/) is a package that allows us to plot interactive 3D scatterplots.
+
+```{r 3d_scatter}
+pca_3d_plot <- plot_ly() %>%
+  add_trace(data = pca_plot_df,
+            x = ~ PC1,
+            y = ~ PC2,
+            z = ~ PC3,
+            mode = "markers",
+            type = "scatter3d",
+            color = ~ molecular_subtype,
+            colors = colorblindr::palette_OkabeIto[1:4])
````
If we are doing this, can we simplify it a bit? I'm not sure whether `add_trace()` is needed based on https://plotly.com/r/3d-scatter-plots/ It seems like we might be able to do this with something like the following, which is maybe more similar to ggplot syntax?
```
pca_3d_plot <- plot_ly(data = pca_plot_df,
                       x = ~ PC1,
                       y = ~ PC2,
                       z = ~ PC3,
                       color = ~ molecular_subtype,
                       colors = colorblindr::palette_OkabeIto[1:4],
                       mode = "markers",  # is this needed? not sure
                       type = "scatter3d"  # same here, does it automatically infer 3d? probably, but explicit is fine
                       ) %>%
  add_markers()
```

jaclyn-taroni replied:
Specifying `mode` and `type` avoids warnings that I thought could be potentially confusing (not sure how `add_markers()` fits in). I had trouble specifying `x`, etc. in the `plot_ly()` call but I think I was missing `~` when I tried that at first. I can take another look. 

This is probably not a good idea, but what about multiple 2D plots _and_ this 3D plot?

jashapiro replied:
TIMTOWTDI

I think `add_markers()` is more like `geom_point()` where `add_trace()` is a more general thing that can do lines too?

jaclyn-taroni commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
````diff
@@ -0,0 +1,351 @@
+
+[`plotly`](https://plotly.com/r/) is a package that allows us to plot interactive 3D scatterplots.
+
+```{r 3d_scatter}
````
Commit messages never lie

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
````diff
@@ -0,0 +1,378 @@
+                                   distance = "pearson")
+```
+
+It seems like there are 3 main stable clusters from the consensus clustering results.
````
I feel like we are missing a bit of explanation of the expected output here. A bit more context about how you concluded 3 clusters might also be useful for people coming back to the notebook.

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
```diff
@@ -0,0 +1,378 @@
+### Silhouette coefficient
+
+The silhouette coefficient or value is a measure of cluster consistency that ranges from -1 to 1.
+It is calculated on a per-sample basis -- it measures how similar a sample is to the cluster it's in compared to other clusters ([Wikipedia entry for silhouette](https://en.wikipedia.org/wiki/Silhouette_(clustering))).
```
To be explict: a value close to 1 means it is very similar to it's cluster and -1 would mean it is very dissimilar?

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
````diff
@@ -0,0 +1,378 @@
+length(sample_index) / length(cc_cluster_labels)
+```
+
+And now we'll filter the RNA-seq matrix.
````
Are we choosing to ignore samples that don't cluster? If so, we should justify this?

jaclyn-taroni replied:
Only for the silhouette value calculations because doing that for a cluster of 1 sample (or 2, for that matter) doesn't make a lot of sense to me. I'm also happy to rip out the silhouette value stuff to save on time to be honest. (I feel like this is a bit of an odd use without enough context.) The main reason for including consensus clustering is _skepticism_ about clustering results.

Thoughts?

jashapiro replied:
That makes sense. 

I do wonder if the silhouette juice is worth the squeeze. If you spend more time on the "consensus because clusters can be janky" I think that makes the rhetorical point.

jaclyn-taroni replied:
:chopped:

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
```diff
@@ -0,0 +1,378 @@
+# Like umap(), prcomp() expects the features to be columns
+pca_results <- prcomp(t(rnaseq_mat[high_variance_index, ]), 
+                      scale = TRUE)
+# The low-dimensional representation is returned in x
```
```suggestion
# The principal component representation is returned in x
```

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
```diff
@@ -0,0 +1,378 @@
+                       x = ~ PC1,
+                       y = ~ PC2,
+                       z = ~ PC3,
+                       mode = "markers",
```
This is redundant with "scatter3d" it seems.
```suggestion
```

jashapiro commented on machine-learning/02-openpbta_consensus_clustering.Rmd:
````diff
@@ -0,0 +1,378 @@
+  ylab("PC3")
+```
+
+`summary()` will report the proportion of variance explained by each principal component.
+By accessing the `importance` element with `<summary results>$importance`, we can use indexing to only look at the first 10 PCs.
````
I think this is a good thing to throw in, but it seems oddly placed here. Maybe some more context about why we care would help? How would you use this information to inform plotting?

jaclyn-taroni commented:
Ready for another 👀  @jashapiro - I added some material from the Monti _et al._ paper that I typically touch on in slides to the notebook (thinking about forgoing slides on consensus clustering in that case 🤔 ) to add more context and removed the silhouette coefficient section.
